### DISCUSSION TITLE
DownloadHelper new version 8.2.0.23 queue malfunction

### DISCUSSION THREAD

**Discussion by JoeKing12001** [aclap-dev/video-downloadhelper#139] [Help and Q&A]
### Your question

This new version is downloading queued files in random order and not the order they were added.  Was some new setting added that I can change to fix this?

### Addon & CoApp version and details

version: 8.2.0.23
channel: stable
build date: 2024-02-14
build options: linuxlic: false, noyt: false, browser: firefox
lang: en-US
license: unset
platform: x86-32 win
UA: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/115.0
coapp: yes. Info:
{
  "id": "net.downloadhelper.coapp",
  "name": "VdhCoApp",
  "version": "2.0.17",
  "binary": "C:\\Program Files\\DownloadHelper CoApp\\vdhcoapp.exe",
  "displayName": "VdhCoApp",
  "description": "Video DownloadHelper companion app",
  "target": {
    "os": "windows",
    "arch": "x86_64",
    "node": "18"
  },
  "home": "C:\\Users\\18282",
  "converterError": "Exit without answer"
}

**Comment by paulrouget** [comment]
> This new version is downloading queued files in random order and not the order they were added. Was some new setting added that I can change to fix this?

Oh, right. My bad. I broke that. I'll work on a fix as soon as I have some time.

Thank you for reporting!

**Reply by dajana91** [reply]
@paulrouget Hi, I have exactly the same problem. I have set that only 1 download should take place at a time, but in the exact order in which files are added to the queue. But Video Downloadhelper downloads totally randomly. The problem is that I automatically rename all downloaded files using a batch tool, with 1_, 2_, 3_ etc. in front of the file names. So that it is exactly in the correct order (by creation date). Due to the random download, none of this works and everything has to be renamed manually. I read the post here today and it is now 3 weeks old. When will the known bug be fixed?

**Reply by dajana91** [reply]
Update: I have tried beta 8.2.2.5a1. Recognising videos via the context menu works better. But the order of adding videos to the queue is still not respected - unfortunately it still happens randomly. I see the correct order, but it just loads some entry in the middle, not in the order shown.

**Reply by dajana91** [reply]
PS: Unfortunately, errors often occur:

moz-extension://XXXXXXXXXXXX/background/main.js:1
receive@moz-extension://XXXXXXXXXXXX/background/main.js:1:13157
id</callCatchAppNotFound/</<@moz-extension://XXXXXXXXXXXX/background/main.js:1:42624

**Comment by JoeKing12001** [comment]
Something else of note.....When there is an error of some sort, it does not indicate which file, by name or URL, that had the error, making it difficult to figure out which thing needs to be re-downloaded.

    On Saturday, February 17, 2024 at 12:13:36 AM EST, Paul ***@***.***> wrote:  
 
 



This new version is downloading queued files in random order and not the order they were added. Was some new setting added that I can change to fix this?


Oh, right. My bad. I broke that. I'll work on a fix as soon as I have some time.

Thank you for reporting!

**Comment by paulrouget** [comment]
I believe I fixed this issue in version [8.2.2.6](https://www.downloadhelper.net/beta) (beta version for firefox). Now downloads are triggered in the order they are added.

@dajana91 @JoeKing12001 Can you please verify this?

**Reply by dajana91** [reply]
[@paulrouget] 

Version 8.2.2.6 has unfortunately not fully fixed the problem. The order of downloading according to the order of adding is not always respected. I am using 8.2.2.6.a1 (beta). The problem still exists. For example, I have a list of 13 videos, each video is on its own page (tab in this case). I go from video page to video page, right click, download video, to the next tab - and quite quickly. Maybe it depends on the speed of adding the links? Back then it wasn't a problem, I could add link by link - click, click, click - and the order was understood and adhered to. Unfortunately not now.

Just as an example: 10 videos. I open tab 1 with video 1, right click, download. Then tab 2 with video 2 etc. up to video 10. Sometimes it happens that the entire list is in the correct order, but suddenly after video 2, video 8 is loaded first and then video 3. Although the entire download list is displayed in the correct order.

I tried to add the links slower, but the same thing happens as you can see here - the correct order in the list, but for whatever reason one of the lower videos gets downloaded first:

![2024-03-10_21-44-50_screenshot](https://github.com/aclap-dev/video-downloadhelper/assets/110555323/903f26ae-5703-4d33-86b5-ab677c2c874b)

**Reply by dajana91** [reply]
So, I tried it out for an hour or so. The bug still exists. Does not matter how fast or slow I am adding the links (of course as long as there are several links in the queue - if you always add a link after the previous one has been downloaded, the test makes no sense).

**Reply by dajana91** [reply]
@paulrouget Still not working... all entries are in order. But for some reason sometimes videos from the middle of the queue or (in this screenshot) of the end are loading next. Makes no sense and unfortunately messes everything up. The problem: After downloading ALL videos, I sort them automatically by batch based on the creation date and then add 1_, 2_, 3_ etc automatically. in front of the file names. Because the videos are loaded in a mess, this sorting is no longer useful.

![2024-03-10_23-58-39_screenshot](https://github.com/aclap-dev/video-downloadhelper/assets/110555323/ec2e2a7a-0527-4fbf-bf69-b314beff0166)

**Reply by dajana91** [reply]
@paulrouget Tried it again today on another site. Still the same issue. Everytime I add a link, a blue circle with a number appears next to the Video Downloader icon (as usual). When I add a second link (when the first one has not already been downloaded), the 1 changes to 2 and it adds up with every new link. And when one video is downloaded, the number decreases (as intended). So I can SEE that a link is successfully added right after my click. 1, 2, 3, 4, 5. At the moment I have to wait everytime after adding a link that the blue circle with the "1" inside disappears, because otherwise the queue is not processed in the correct order. That was not the case in the past and very annoying :-( If I want to add for example 7 links to download videos there should be a 7 in the blue circle (or maybe less, because some videos might have already been downloaded during adding links). But that makes no sense, because it is not downloaded in the added order. As I wrote: Latest beta-version 8.2.2.6a1. Does NOT work.

**Reply by paulrouget** [reply]
@dajana91 I hear you. I'm looking at it. I'll let you know once I figured it out.

**Reply by paulrouget** [reply]
@dajana91 I have difficulties reproducing here. But I made a change in version 8.2.2.7 which might improve your situation. Can you give it a try for me please? (you just need to update the addon, or click that link: https://www.downloadhelper.net/data/vdh-mozilla-beta-8.2.2.7a1.xpi )

**Reply by paulrouget** [reply]
Appreciated. I'm actively trying to figure this one out. Thanks a lot for testing and for your patience.

**Reply by dajana91** [reply]
@paulrouget Of course, this is constructive mutual help. I would be really pleased if it worked again soon.

**Reply by paulrouget** [reply]
@dajana91 in the video, before downloading it, can you go in the menu, click Details, then open the `core_media` section, and tell me what's written next to `builder`?

Something like this:

<img width="626" alt="Screenshot 2024-03-14 at 11 32 57" src="https://github.com/aclap-dev/video-downloadhelper/assets/373579/352dac58-e9fe-45a5-ae83-8a6c0ab374b6">

**Reply by dajana91** [reply]
@paulrouget Sure, here are the details:

![2024-03-14_23-01-12_screenshot](https://github.com/aclap-dev/video-downloadhelper/assets/110555323/60040d47-636a-465c-a1cf-553dd8a7a88c)

**Reply by dajana91** [reply]
@paulrouget Did my information help with troubleshooting? Does the information give any clues as to what the problem may be?

**Reply by paulrouget** [reply]
Yes, absolutely. I understand why I could not reproduce. Thank you, still looking into it.

**Reply by dajana91** [reply]
@paulrouget Are there any news? The download order is still not respected. And I have deliberately set that only 1 file should be downloaded at a time so that the download sequence is consciously respected. But it still doesn't work.

**Reply by dajana91** [reply]
@paulrouget Even 8.2.2.8a1 has NOT solved the problem. As before, the order in which links are added when downloading is NOT respected.

**Reply by Mattmjs** [reply]
@dajana91 , have you tried using the browser to do the downloading. The screenshot you posted above indicate that they
are basic mp4 files which the browser could handle.

**Reply by dajana91** [reply]
Ahm ... Yes? I know what I'm doing 😂 These are not "just click and download" MP4 files. I use the Video Downloader for this, which also renames the files according to the RegEx specifications. This is a bug that never used to occur. But unfortunately, no more feedback seems to be coming back.

**Reply by dajana91** [reply]
No updates yet? Nothing new about a possible fix? @paulrouget

**Reply by dajana91** [reply]
@paulrouget Hello?

**Comment by dajana91** [comment]
@paulrouget Well, I am using 8.2.2.5a1 since yesterday and the problem still exists (Firefox), tried it out several times.

**Reply by GrampaWildWilly** [reply]
You are not running the beta version that Paul identified.  You need to use the one he mentioned.

**Reply by dajana91** [reply]
@GrampaWildWilly Yes, if the link from Paul led to 8.2.2.6, I would do that. But the link leads to 8.2.2.5. 🤷‍♂️

**Comment by dajana91** [comment]
So where do I get the 8.2.2.6 version? As noted, the link only leads to 8.2.2.5a1

**Reply by Mattmjs** [reply]
You can force the update, **Ctrl+Shift+A** to go to extensions page. There is a gear icon, check for updates.

**Reply by GrampaWildWilly** [reply]
That's a pretty easy way of doing it.  There's another easy way.  Click the Wiki link at the top of this page.  That takes you to a user manual of sorts.  You should get in the habit of using that frequently as a reference.  There is a Table of Contents down the right side of the web page.  There's plenty of good content there that you should read.  Relevant to this discussion, click the chapter about Beta versions.  I think you'll be able to take it from there.

**Reply by dajana91** [reply]
@GrampaWildWilly I know this beta page. But there is only version 8.2.2.5 and NOT 8.2.2.6 available. 8.2.2.5 still has the bug, but @paulrouget has pointed to version 8.2.2.6, which is not downloadable there. But the tip from @GrampaWildWilly was helpful.

**Reply by GrampaWildWilly** [reply]
It seems there is a technical problem on GitHub with the link to the beta within the Wiki.  The technique of using Firefox's own Check for Updates button (as @Mattmjs described) appears to be the only way this is working for the moment.  The Check for Updates button does not appear until you switch off automatic updates for VDH.  You should switch off automatic updates for VDH.  There is an issue with Firefox's automatic updater.  If Firefox decides it's time to do an automatic update to VDH (or any other extension), it does it.  Now.  Including if VDH happens to be in the middle of actually downloading something.  So Firefox will cause your VDH download to abnormally & abruptly terminate, leaving you with an unplayable file.  Your only recourse is to start the download over from the beginning.  So you should turn off automatic updates for VDH, both general availability & beta versions, until this automatic update issue is addressed.

**Comment by fernando-sc** [comment]
I'm having this problem too. Very annoying, because I'm downloading a course to watch later in vlc, and I rename the files sorting by time of download conclusion, even limited the simultaneous download to this.

Is it solved? How can I use the beta version?

**Reply by jcfrog** [reply]
hi, 
the beta versions are available from here: https://github.com/aclap-dev/video-downloadhelper/wiki/Beta-versions

**Reply by dajana91** [reply]
@fernando-sc I am doing exactly the same: loading courses and the problem still exists. The download order is not respected.

**Comment by JoeKing12001** [comment]
Could you make it say the name of the file that had the error?All it says is....
JSON.parse: end of data after property value in object at line 115 column 6 of the JSON data
moz-extension://28f6edad-7b60-43cf-bfe3-caf29ad9a805/background/main.js:42
***@***.***://28f6edad-7b60-43cf-bfe3-caf29ad9a805/background/main.js:42:20393

And I have no way of figuring out which thing to re-download.
Also, download order seems to be fixed now, and I just wanted to say thanks.  I'm sure you don't hear it enough.

**Reply by dajana91** [reply]
@JoeKing12001 No, it has not been fixed (8.2.2.8a1 beta). The order is still not taken into account. @paulrouget has not replied for a long time or written anything about the status of troubleshooting. 🤷‍♂️ I would like to thank you, but the problem exists and there is no more feedback.

**Comment by JoeKing12001** [comment]
I have noticed that again, but the time before last, all files went in the correct order, which would mean the problem is somehow intermittent.

    On Thursday, April 11, 2024 at 06:34:13 PM EDT, dajana91 ***@***.***> wrote:  
 
 


@JoeKing12001 No, it has not been fixed (8.2.2.8a1 beta). The order is still not taken into account. @paulrouget has not replied for a long time or written anything about the status of troubleshooting. 🤷‍♂️ I would like to thank you, but the problem exists and there is no more feedback.

**Reply by dajana91** [reply]
I find it critical that there is no longer any response at all.

**Reply by paulrouget** [reply]
Sorry, I've been busy with some other things.

I haven't been able to reproduce this yet. We have a new version coming up very soon (that's why I was busy) which changes the queuing system quite extensively. I hope this will address your issue.

**Comment by JoeKing12001** [comment]
Well, I have a solution everyone can use in the meantime, but it may be impractical for some people.  If you set the number of simultaneous downloads to a very large number, they all start as soon as you add them and therefore sequential.

    On Monday, April 15, 2024 at 02:44:47 AM EDT, Paul ***@***.***> wrote:  
 
 


Sorry, I've been busy with some other things.

I haven't been able to reproduce this yet. We have a new version coming up very soon (that's why I was busy) which changes the queuing system quite extensively. I hope this will address your issue.

**Comment by fernando-sc** [comment]
Lmao. I did exactly the opposite. I've set the max simultaneous downloads
to just one. It's ridiculous, but been working, just getting much more time
than necessary.

Dev, pls never forget the golden rule: don't fix what's not broken.

Let's wait the new version, then.

Em dom., 21 de abr. de 2024, 09:50, JoeKing12001 ***@***.***>
escreveu:

> Well, I have a solution everyone can use in the meantime, but it may be
> impractical for some people.  If you set the number of simultaneous
> downloads to a very large number, they all start as soon as you add them
> and therefore sequential.
>

**Comment by dajana91** [comment]
Still not working. Worked in the past. Don´t know why they changed it to the bad. To me (personally) no longer useable. Sad.

**Reply by paulrouget** [reply]
Do you use "Download As" or regular "Download"?

**Comment by JoeKing12001** [comment]
Usually, just download, especially with bulk quantities.
    On Tuesday, July 9, 2024 at 05:47:28 AM EDT, Paul ***@***.***> wrote:  
 
 


Do you use "Download As" or regular "Download"?

**Comment by GrampaWildWilly** [comment]
No, the queuing mechanism still does not launch the downloads in the order in which I request them.  Here is exactly what I did, running Windows 11 Pro 64-bit version 23H2 build 22631.3958, Firefox 128.0.3 64-bit, licensed VDH 9.1.0.34a1 beta, CoApp 2.0.19.

- I set my `Max concurrent downloads` to 1.  This guarantees that as I request downloads, they queue up instead of all running concurrently.
- I opened these URLs in tabs in the same browser window:
https://www.youtube.com/watch?v=DYwQy_9JPtQ
https://www.youtube.com/watch?v=mTUhnIY3oRM
https://www.youtube.com/watch?v=-KtYgOM62fk
https://www.youtube.com/watch?v=jMcWldfg28s
https://www.youtube.com/watch?v=RStkIO7JQd0
https://www.youtube.com/watch?v=3rOer5xMFM4
https://www.youtube.com/watch?v=0Y5b0fg_ZwE
https://www.youtube.com/watch?v=J36z7AnhvOM
https://www.youtube.com/watch?v=DAgQIb77UhU
https://www.youtube.com/watch?v=LcJm1pOswfM
The URLs actually don't matter.  They may as well be random.  But for the sake of full disclosure, these are the ones I used.
- On the first URL, I clicked the `Download As...` function.  This opened a file selection dialog.
- In the file picker, after navigating to the selected directory, I modified the proposed file name by prefixing it with a number, starting with `#01`.  This was so I could track the order in which I launched my downloads.
- I did not launch the download at this point.  I just left the file picker waiting for me to hit Enter.
- I went to each browser tab in turn & did the same steps to initiate `Download As...`, modify the proposed file name by prefixing a number, then without actually launching the download, go to the next browser tab.
- I used Windows task switching to stack up the 10 file picker dialogs so they were in numerical order.
- I started on `#01` & hit Enter.  Then `#02`.  Then `#03`.  And so on until I reached `#10`.  This guarantees that I launched the downloads in numerical order.

The downloads did not occur in numerical order.  Here is an image of the directory into which I downloaded these videos.

![#01](https://github.com/user-attachments/assets/3546ad31-b90d-4096-85c9-c5f9ba52b2c4)

Look closely at the column `Date modified`.  You can see that this directory is sorted in descending order by the modified date/time of the file.  That is, the most recent file is at the top & the oldest file is at the bottom.  So the first downloaded file is at the bottom & the order in which they were downloaded is bottom to top.  You can see from the sequence numbers I prefixed to the file names that these files are NOT in numerical order.

You'll have to take my word for it that the notifications I got from Windows as each download completed were in this same order.

Here is the history listing of what VDH did here.

![#02](https://github.com/user-attachments/assets/4eec4d4e-4c02-4179-a073-7854db59e336)

This is not a screenshot via the PrintScreen key.  This is a screenshot using Firefox's own screenshot function.  I did this in order to capture the entire history, which would need to be scrolled because it doesn't fit in a single screen.  You'll note that the order in which VDH itself says these downloads occurred was the same as what you can see in the download directory in the first image.

This is not working.  Some of these downloads got picked off the queue out of order.  This problem is not fixed.

**Reply by GrampaWildWilly** [reply]
My example is made up.  I encourage you folks who originally complained about this to try this with your real-world videos.  Is this fixed now?  Please post your results using this new VDH beta.

**Reply by paulrouget** [reply]
To be clear:

- this issue doesn't exist if you use just "Download", correct?
- the issue is that they don't **start** in the order the blue download buttons were clicked on, correct?

> I did not launch the download at this point. I just left the file picker waiting for me to hit Enter.

- Does this problem go away if, instead of waiting, you don't leave the file pickers hanging and just enter the name and validate immediately. As in, instead of doing: DownloadAs, DownloadAs, DownloadAs, ValidateFilePicker, ValidateFilePicker, ValidateFilePicker, you do: DownloadAs, ValidateFilePicker, DownloadAs, ValidateFilePicker, DownloadAs, ValidateFilePicker, then there's no problem, correct?

The "date modified", and the order in the history are based on the time the download is *finished*. So unrelated.

**Reply by GrampaWildWilly** [reply]
There's some comments above about setting `Max concurrent downloads` to either 1 or a large number.  Setting it to 1 is the only way I know of to guarantee that I will get a queue.  Furthermore, with single-threaded downloads, the first one completes, then the second one starts, & the third one doesn't start until the second one completes.  (At least, that's how it will work once @paulrouget Paul really does fix this.)

If you let VDH run multiple concurrent downloads, they will not complete in the order in which you launched them.  They also will not necessarily complete shortest one first, longest one last.  You would think they would complete in their order of length but my experience is that the bandwidth allocated to each download is not equal.  Some get more bandwidth, some less.  Sometimes, a large file gets a lot of bandwidth & completes before a shorter file which gets little bandwidth.  The allocation of bandwidth to multiple concurrent downloads is beyond my ability to predict.  If there's an algorithm to the allocation of bandwidth to multiple concurrent downloads, I haven't discovered it.

The only way to make the process deterministic is to single-thread the downloads.  On the other hand, if you're interested only in the order in which they launch & not the order in which they finish, then single-threading is of no consequence to you.

Truth is, since bandwidth is a finite resource, there is really not a lot of advantage in concurrent downloading.  There might be up to a point, but there is a point of diminishing returns.  The collection of downloads you might be executing shares your bandwidth.  They must all fit within whatever you are paying your ISP for.  If you get 10 downloads running at the same time, each one will actually download quite slow.  Or, as I just observed, some will run fast & some slow, but there's no predicting which will get lucky & which will bring up the rear.  In the end, the total cumulative time to download all your files will probably be the same single-threaded or multi-threaded.  I have found that in practice, I shouldn't try to download more than 3 or 4 files at any given moment.  If I'm downloading let's say 10 or 15 files, I have tools that launch 3 or 4 at a time, wait for those to complete, then launch the next few downloads.  As each download in a group completes, the others absorb the bandwidth being freed up & they speed up.  Like I say, there's a point at which it just doesn't pay to try to do more downloads.  You need to carefully observe your environment & make a guess at what your practical maximum concurrency should be.

**Comment by GrampaWildWilly** [comment]
> The "date modified", and the order in the history are based on the time the download is finished. So unrelated.

That is true in the general sense, but since I am single-threading my downloads, it's not true.

**Reply by GrampaWildWilly** [reply]
Let me rephrase that.  In this test the order in which I start the downloads is the order in which they should finish because I have `Max concurrent downloads` set to 1.  The start order should be the end order.  I'm single-threading.  Download n should complete & only after that should download n+1 start.  The order of finishing is the same as the order of starting because `Max concurrent downloads` = 1.

**Comment by GrampaWildWilly** [comment]
Just clicking `Download As...` & launching the download via the file picker right away does not exercise this problem properly.  By the time I get to the second tab, the first download completes.  I'm not creating a queue.  The point of this exercise is to create a queue.  I need to load up enough downloads that there is a backlog in the queue.  To properly do what you're describing, I would have to find a bunch of long downloads so that I could launch the first one & it would still be running by the time I complete the request for the second one.  And I'm not doing that.  I refuse to waste that kind of bandwidth.  The point here is to do something that is easy enough to cause a queue to accumulate without spending hours running the test.  You need to waste those hours because you're the developer.  I'm a user.  I'm not going to do it.  My test, with VDH set to single-thread -- that's important, I'm single-threading VDH -- is valid.  You need to try my test.  I don't need to try a test with large files.  You do.

Here's what I think is happening.  It's a guess.  I'm trying to fit an explanation to what happened.  The first download was `#01` but the second download was `#05`.  `#05` was in the middle of the sequence I hit Enter on.  How could it be that with at least 3 other downloads already in the queue -- numbers `#02`, `#03`, & `#04` -- VDH completed `#01` & chose `#05` to download next, in preference to `#02`, which was supposedly at the head of the queue.  I think it just happened that `#01` completed about the time I hit Enter on `#05`.  That caught VDH's attention & distracted it.  VDH went to that download for which I just hit Enter instead of putting `#05` at the back of the queue & actually picking `#02` off the head of the queue.

But there was another download done out of sequence.  `#05` wasn't the only one out of sequence.  I'm pretty sure that by the time VDH had downloaded `#01` & `#05`, I had hit Enter on all 10 file pickers.  Why did `#10` follow `#05`?  I can't even make up a far-fetched explanation for that.  The obvious explanation is that hitting Enter does not ALWAYS put that request at the back of the queue.  Another obvious explanation is that VDH does not ALWAYS go to the queue to select its next download.  Both of those should always be the case but it looks like they're not.  You've got the code.  You need to look for the place or places where VDH decides what to do next.  I suspect you've got fancy logic that allows it to make multiple choices when it should be making just one choice.  Internally, VDH should queue EVERY request.  It should not try to optimize certain situations to try to get things done faster.  VDH should slavishly always place a request at the back of the queue.  VDH should also slavishly always select the front entry in the queue as the next thing to download.  Always use the queue.  Make it FIFO.  No alternative logic paths in an attempt to speed things up.  In this discussion people are saying deterministic behavior is the only thing that is important.  We're not getting that.

When I talk about a queue, I am not talking about the queue that backs up because I've set VDH to single-thread.  I'm talking about an internal data structure.  VDH should always use this internal FIFO queue for scheduling, no matter what level of multi-threading is set by the `Max concurrent downloads` setting.  If you already have the maximum allowed number of concurrent downloads, don't pick off the head of queue yet.  Wait until a download completes, then do that.  Just always use this internal queue.  Never avoid it.  Even if you put something on the queue & immediately take it off.  Don't avoid it.  Make it perform deterministically.

**Reply by paulrouget** [reply]
I think it's an issue with the filepicker messing up the order.

Can someone simply answer this question (if you don't know, no need to say you don't know): does this bug exists without using "Download As" but just "Download"?

**Comment by JoeKing12001** [comment]
Can not download from this......https://hdbest.net/47318-the-witches-of-breastwick-video-2005.html, could you please experiment with it and include a method to download blob: in your next update/version?  I've noticed that any site that uses blob: protocol can't be downloaded.

    On Monday, April 15, 2024 at 02:44:47 AM EDT, Paul ***@***.***> wrote:   

 


Sorry, I've been busy with some other things.

I haven't been able to reproduce this yet. We have a new version coming up very soon (that's why I was busy) which changes the queuing system quite extensively. I hope this will address your issue.
